File: src/Coachable/sorting/QuickSortEntropy.java
package Coachable.sorting;

public class QuickSortEntropy {
	private QuickSortEntropy() {
	}

	public static void sort(Comparable[] arr) {
		sort(arr, 0, arr.length - 1);
	}

	private static void sort(Comparable[] arr, int low, int high) {
		if (low >= high) return;

		Comparable pivot = arr[low];
		int i = low, lt = low, gt = high;

		while (i <= gt) {
			if (arr[i].compareTo(pivot) < 0)
				swap(arr, i++, lt++);
			else if (arr[i].compareTo(pivot) > 0)
				swap(arr, i, gt--);
			else i++;
		}
		sort(arr, low, lt - 1);
		sort(arr, gt + 1, high);
	}

	private static void swap(Comparable[] arr, int i, int j) {
		Comparable temp = arr[j];
		arr[j] = arr[i];
		arr[i] = temp;
	}
}

File: src/LearningMaterial/1-DSA/4-String,Sorting/QuickSortEntropy.py
def quick_sort(arr, lo, hi):
    if lo >= hi: return

    lt, gt = lo, hi
    i = lo + 1
    pivot = arr[lo]

    while i <= gt:
        curr = arr[i]
        if curr < pivot:
            arr[lt], arr[i] = arr[i], arr[lt]
            i += 1
            lt += 1
        elif curr > pivot:
            arr[gt], arr[i] = arr[i], arr[gt]
            gt -= 1
        else:
            i += 1

    print("now arr=", arr)
    quick_sort(arr, lo, lt - 1)
    quick_sort(arr, gt + 1, hi)


def sort(arr):
    quick_sort(arr, 0, len(arr) - 1)
    return arr


arr = [3, 8, 2, 15, 27, 21, 17, 10, 16, 7, 24, 0, 4, 6, 18, 5]

sort(arr)
print(arr)
